generator/src/main/java/org/apache/kafka/message/checker/Unifier.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kafka.message.checker;
+
+import org.apache.kafka.message.EntityType;
+import org.apache.kafka.message.FieldSpec;
+import org.apache.kafka.message.FieldType;
+import org.apache.kafka.message.MessageGenerator;
+import org.apache.kafka.message.MessageSpec;
+import org.apache.kafka.message.StructRegistry;
+import org.apache.kafka.message.StructSpec;
+import org.apache.kafka.message.Versions;
+
+import java.util.List;
+
+import static org.apache.kafka.message.checker.CheckerUtils.max;
+import static org.apache.kafka.message.checker.CheckerUtils.min;
+
+/**
+ * The unifier attempts to unify the types of two KRPC messages. In other words, to see them as
+ * two different points in the evolution of a single message
+ * specification.
+ */
+class Unifier {
+    private final MessageSpec topLevelMessage1;
+    private final StructRegistry structRegistry1;
+    private final MessageSpec topLevelMessage2;
+    private final StructRegistry structRegistry2;
+
+    Unifier(
+        MessageSpec topLevelMessage1,
+        MessageSpec topLevelMessage2
+    ) throws Exception {
+        this.topLevelMessage1 = topLevelMessage1;
+        this.structRegistry1 = new StructRegistry();
+        this.structRegistry1.register(topLevelMessage1);
+        this.topLevelMessage2 = topLevelMessage2;
+        this.structRegistry2 = new StructRegistry();
+        this.structRegistry1.register(topLevelMessage2);
+    }
+
+    static FieldSpec structSpecToFieldSpec(StructSpec structSpec) {
+        return new FieldSpec(structSpec.name(),
```

Comment:
These constructor calls are not pretty. Would it be too weird to have a build _and_ the jackson constructor?

Reply:
Yeah, it is a bit messy. I think it might be hard to keep things in sync if I added builders for these, though... since we have to support how Jackson builds them.

Reply:
Actually, it looks like Jackson does support builders. `@JsonPOJOBuilder(prefix="set", buildMethodName="build")`

We don't need to add this here, but maybe something to consider

Reply:
That might be better than what we have. I wonder if there's any runtime performance hit? (Doesn't matter for the generator much, but it does matter for things like our JSON handling inside the server)...